pkgs/build-support/php/build-php.nix

```diff
@@ -0,0 +1,106 @@
+{ stdenvNoCC, lib, writeTextDir, composer, php }@inputs:
+
+{ pname
+, version
+, src
+, composerLock ? "__unset"
+, vendorDir ? "vendor"
+, vendorHash
+, php ? inputs.php
+, composer ? inputs.composer
+, ...
+}@args:
+
+let
+  vendor = stdenvNoCC.mkDerivation (finalAttrs: {
+    inherit src version;
+
+    pname = "${pname}-vendor";
+
+    nativeBuildInputs = [
+      composer
+    ];
+
+    strictDeps = true;
+
+    configurePhase = ''
+      runHook preConfigure
+
+      if [ -f ${src}/composer.lock ]
+      then
+        cp -f ${src}/composer.lock composer.lock
+      fi
+
+      if [ -f ${composerLock} ]
+      then
+        cp -f ${composerLock} composer.lock
+      fi
+
+      if [ ! -f "composer.lock" ]
+      then
+        echo "No composer.lock file found"
+        exit 1
+      fi
+
+      runHook postConfigure
+    '';
+
+    buildPhase = let
+      # See https://getcomposer.org/doc/03-cli.md#composer-home-config-json
+      composerHome = let
+        content = {
+          config = {
+            autoloader-suffix = "${builtins.replaceStrings ["-"] [""] pname}";
+          };
+        };
+      in
+        writeTextDir "/config.json" (builtins.toJSON content);
+    in ''
+      runHook preBuild
+
+      COMPOSER_HOME=${composerHome} \
+      COMPOSER_VENDOR_DIR=${vendorDir} \
+      composer install \
+        --no-cache \
+        --no-dev \
+        --no-interaction \
+        --no-scripts \
+        --optimize-autoloader \
+        --quiet
+
+      runHook postBuild
+    '';
+
+    installPhase = ''
+      runHook preInstall
+
+      mkdir -p $out
+      cp -ar ${vendorDir} $out/
+
+      runHook postInstall
+    '';
+
+    outputHashMode = "recursive";
```

Comment:
Just stumble upon `composer install --download-only` in the documentation that could to alleviate some of the issues I have mentioned before. Instead of trying to download/extract the dependencies and do the code generation in the same derivation it could be split in two: first the cache is primed using `composer install --download` using a FOD then the code generation is done in another derivation.
We still want to make it reproducible so the random ID that Composer might generate needs to be dealt with. It should however be less likely to require to update all the FOD when updating Composer.

Reply:
Hmm, I guess we would have to remove the autoloader part. But I guess at that point we would want to go the repo route like [composition-c4 does](https://github.com/fossar/composition-c4/blob/493ddc4a8279cde3bee0db1d04fd7d4838d51ab9/src/composer-setup-hook.sh#L20).

Reply:
> Composer does not guarantee stability of what it outputs in `vendor/`,

Apparently it does on some conditions:

[![php composer reproducibility matrix](https://planttext.com/api/plantuml/png/lPB1JWCn34Jl-Of8t29HVg2gAlN2gQguTdj8DeQQJUp8cmNwztYtbAeIz820OqRpF1FBoxvwgHafPb-CwJiaAzo3CSvkxZSDhJWNVKj30aKu3Yq6J5WF3RrYtTjvMBGV5ekaqrAQtuL5GwfcJr47EVi0Qi3AefG24tYYPedRRaJKwCxgSyzqDlkXSc8VGMwBm2E-dG3tNG4ROEzVm19BLkamcTtsvaa0Q3SscZxZGu8hmJpKNymMY3kxvchbClsh_Op9_xIYX1rSU8OIVOLF_ahplzdRZmkP9UWX7u_v7G00)](https://planttext.com/?text=lPB1JWCn34Jl-Of8t29HVg2gAlN2gQguTdj8DeQQJUp8cmNwztYtbAeIz820OqRpF1FBoxvwgHafPb-CwJiaAzo3CSvkxZSDhJWNVKj30aKu3Yq6J5WF3RrYtTjvMBGV5ekaqrAQtuL5GwfcJr47EVi0Qi3AefG24tYYPedRRaJKwCxgSyzqDlkXSc8VGMwBm2E-dG3tNG4ROEzVm19BLkamcTtsvaa0Q3SscZxZGu8hmJpKNymMY3kxvchbClsh_Op9_xIYX1rSU8OIVOLF_ahplzdRZmkP9UWX7u_v7G00)

Reply:
I just added `--apcu-autoloader-prefix=` so the issue with apcu is fixed as well.

Reply:
> Apparently it does on some conditions:

This is only for a specific version, there is no guarantee across different Composer version.

> I just added `--apcu-autoloader-prefix=` so the issue with apcu is fixed as well.

According to the Composer documentation it implicitly enables `--apcu-autoloader`, not sure if it something we want. We probably want to make sure the APCu loader is never enabled in the `composer.json`. We can take advantages of the level 2 optimization using `--classmap-authoritative` since anyway no class can be added at runtime.

Reply:
> Hmm, I guess we would have to remove the autoloader part. But I guess at that point it would be simpler to go the repo route like [composition-c4 does](https://github.com/fossar/composition-c4/blob/493ddc4a8279cde3bee0db1d04fd7d4838d51ab9/src/composer-setup-hook.sh#L20).

Yep that's more or less the idea I guess, it will be done in 2 phases:
1. FOD: `composer install --download-only` to retrieve the cache (and possibly do some cleanup for deps retrieved from VCS repositories)
2. `COMPOSER_DISABLE_NETWORK=1 composer install ...` using the cache we retrieved at the first step

Reply:
Feel free to submit changes in the repo as well, I won't have the time to work on this until tonight.

Reply:
The second install should be in the non-FOD.

Reply:
Job done, it works :)